src/br/com/zup/nossacasacodigo/cart/Cart.java

```diff
@@ -0,0 +1,48 @@
+package br.com.zup.nossacasacodigo.cart;
+
+import java.math.BigDecimal;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+
+import br.com.zup.nossacasacodigo.book.Book;
+
+public class Cart {
+	private Set<CartItem> items = new HashSet<>();
+	private static BigDecimal finalValue = BigDecimal.ZERO;
+
+	public void addToCart(Book book, int quantity) {
+		if (quantity < 0 || book == null) {
+			throw new IllegalArgumentException("Livro ou quantidade inv�lida");
+		}
+		Optional<CartItem> alreadyInCart = checkIfIsInCart(book);
+		CartItem item;
+		if (alreadyInCart.isPresent()) {
+			item = alreadyInCart.get();
+			finalValue = finalValue.subtract(item.calculateSubtotal());
+			item.setQuantity(item.getQuantity() + quantity);
+			finalValue = finalValue.add(item.calculateSubtotal());
+		} else {
+			item = new CartItem(book, quantity);
+			items.add(item);
+			finalValue = finalValue.add(item.calculateSubtotal());
+		}
+	}
+
+	private Optional<CartItem> checkIfIsInCart(Book book) {
+		for (CartItem cartItem : items) {
+			if (book.compareTo(cartItem.getBook()) == 0) {
+				return Optional.ofNullable(cartItem);
```

Comment:
Nesse ponto, a gente tem certeza que `cartItem` não é nulo, então, se você quiser, pode usar o `Optional.of()` (Eu, particularmente, acho muito ok `Optional.ofNullable()`, uso ele sempre haha)

Reply:
Legal ! Neste caso teria algum problema em manter o Optional.ofNullable() ?

Reply:
Nenhum problema! Inclusive, opinião minha haha, acho que é uma falha de design ter um `Optional.ofNullable` e um `Optional.of` haha

Reply:
Uma outra possibilidade de código aqui seria ```items.stream().filter(item -> item.fromBook(book)).findAny()```